components/net/filemanager_thread.rs

```diff
@@ -108,69 +108,161 @@ enum FileImpl {
                             send_error(load_data.url.clone(), format_err, consumer);
                         }
                         Some((id, _fragment)) => {
-                            self.process_request(&load_data, consumer, &RelativePos::full_range(), &id);
+                            self.process_request(load_data, consumer, RelativePos::full_range(), id);
                         }
                     }
                 },
                 FileManagerThreadMsg::DecRef(id, origin) => {
                     if let Ok(id) = Uuid::parse_str(&id.0) {
-                        self.dec_ref(id, origin);
+                        spawn_named("dec ref".to_owned(), move || {
+                            store.write().unwrap().dec_ref(id, origin);
+                        })
                     }
                 }
                 FileManagerThreadMsg::IncRef(id, origin) => {
                     if let Ok(id) = Uuid::parse_str(&id.0) {
-                        self.inc_ref(id, origin);
+                        spawn_named("inc ref".to_owned(), move || {
+                            store.read().unwrap().inc_ref(id, origin);
+                        })
                     }
                 }
                 FileManagerThreadMsg::Exit => break,
             };
         }
     }
 
-    fn inc_ref(&mut self, id: Uuid, origin_in: FileOrigin) {
-        match self.store.get(&id) {
+    fn process_request(&self, load_data: LoadData, consumer: LoadConsumer,
+                       rel_pos: RelativePos, id: Uuid) {
+        let origin_in = load_data.url.origin().unicode_serialization();
+        match self.store.read().unwrap().entries.get(&id) {
+            Some(entry) => {
+                match entry.file_impl {
+                    FileImpl::Memory(ref buffered) => {
+                        if *entry.origin == origin_in {
+                            let buffered = buffered.clone();
+                            let classifier = self.classifier.clone();
+                            spawn_named("load blob".to_owned(), move ||
+                                load_blob(load_data, consumer, classifier,
+                                          None, rel_pos, buffered));
+                        } else {
+                            let e = format!("Invalid blob URL origin {:?}", origin_in);
+                            send_error(load_data.url.clone(), NetworkError::Internal(e), consumer);
+                        }
+                    },
+                    FileImpl::PathOnly(ref filepath) => {
+                        let opt_filename = filepath.file_name()
+                                               .and_then(|osstr| osstr.to_str())
+                                               .map(|s| s.to_string());
+
+                        if *entry.origin == origin_in {
+                            let mut bytes = vec![];
+                            let mut handler = File::open(filepath).unwrap();
```

Comment:
This should similarly be handled outside the lock.

The locked section perhaps should simply clone the entry, and then something out of it can handle this match.